src/linker/Linker/LinkContext.cs

```diff
@@ -143,7 +143,7 @@ public bool IgnoreUnresolved
 
 		public bool LogMessages { get; set; }
 
-		public ILogger Logger { get; set; } = new ConsoleLogger ();
+		public ILogger Logger { private get; set; } = new ConsoleLogger ();
```

Comment:
set-only properties are generally frowned upon:

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/property

> ❌ DO NOT provide set-only properties or properties with the setter having broader accessibility than the getter.

Reply:
I can definitely change it to a SetLogger method. I just wanted to avoid breaking people with this. I noticed that the error this started with has a custom logger `Xamarin.Android.Tasks.LinkAssemblies.LogMessage`. Would you know how is that logger registered?

Reply:
> Would you know how is that logger registered?

The [`<LinkAssemblies/>` MSBuild task implements the `ILogger` interface](https://github.com/xamarin/xamarin-android/blob/cf6d51e4531ace51188d7920e2be1ef47ad9197c/src/Xamarin.Android.Build.Tasks/Tasks/LinkAssemblies.cs#L24):

```csharp
public class LinkAssemblies : AndroidTask, Mono.Linker.ILogger
```

The [`Linker.Process()`](https://github.com/xamarin/xamarin-android/blob/cf6d51e4531ace51188d7920e2be1ef47ad9197c/src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/Linker.cs#L18-L31) is invoked, passing an instance of [`LinkAssemblies` as the `ILogger` parameter](https://github.com/xamarin/xamarin-android/blob/cf6d51e4531ace51188d7920e2be1ef47ad9197c/src/Xamarin.Android.Build.Tasks/Tasks/LinkAssemblies.cs#L126):

```csharp
Linker.Process (options, this, out link_context);
```

`Linker.Process()`, meanwhile, uses the `ILogger` value to set the [`AndroidLinkContext.Logger` property](https://github.com/xamarin/xamarin-android/blob/cf6d51e4531ace51188d7920e2be1ef47ad9197c/src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/AndroidLinkContext.cs#L6), and `AndroidLinkContext` IS-A `LinkContext`, i.e. it's setting `LinkContext.Logger`:

https://github.com/mono/linker/blob/0c63b9ed74130695ef270002d9d9c572201da90c/src/linker/Linker/LinkContext.cs#L146

...which is the property that I was commenting on here...

Removing this property setter ("rename to `SetLogger()`") would thus result in a compatibility change, breaking the xamarin-android build.

All of this *does* show a problem on the xamarin-android side: https://github.com/xamarin/xamarin-android/blob/master/src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/FixAbstractMethodsStep.cs#L29

```csharp
CheckAppDomainUsageUnconditional (assembly, (string msg) => Context.Logger.LogMessage (MessageImportance.High, msg));
```

Here we have the ~same issue I was originally complaining about -- though it's in *my* repo, so I should go fix that now -- as the `Context.Logger.LogMessge()` invocation is "really" `Context.Logger.LogMessage(MessageImportance.High, msg, new object[0])`, so if `msg` contains any `{` characters, things will still blow up.

Reply:
xamarin-android's `CheckAppDomainUsageUnconditional()` usage fixed in:

https://github.com/xamarin/xamarin-android/commit/c683788921265f6a4ba22b2af71228bea48b05fb